src/test/java/org/jenkinsci/plugins/workflow/cps/steps/RestartingLoadStepTest.java

```diff
@@ -218,4 +226,134 @@ public void pauseInsideLoad() throws Exception {
+            r.jenkins.getWorkspaceFor(p).child("b.groovy").write(
+                    "def m(arg) {echo (/${this} binding=${binding.variables}/); a(/${arg} from b/)}; this", null);
+            p.setDefinition(new CpsFlowDefinition(
+                    EXISTING_VAR_NAME + " = " + conflictingValue + ";" +
+                    "a = 0;" +
+                    "node {a = load 'a.groovy'};" +
+                    "echo (/Pre-semaphore value is ${" + EXISTING_VAR_NAME + "}/);" +
+                    "semaphore 'wait';" +
+                    "echo (/Post-semaphore value is ${" + EXISTING_VAR_NAME + "}/);" +
+                    "def b;" +
+                    "node {b = load 'b.groovy'};" +
+                    "echo (/${this} binding=${binding.variables}/);" +
+                    "b.m('value');" +
+                    "echo (/final injected value is ${" + EXISTING_VAR_NAME + "}/);", true));
+            WorkflowRun b = p.scheduleBuild2(0).getStartCondition().get();
+            SemaphoreStep.waitForStart("wait/1", b);
+            r.assertLogContains("Pre-semaphore value is " + conflictingValue, b);
+        });
+        story.then( r -> {
+            WorkflowJob p = r.jenkins.getItemByFullName("p", WorkflowJob.class);
+            WorkflowRun b = p.getBuildByNumber(1);
+            SemaphoreStep.success("wait/1", null);
+            r.waitForCompletion(b);
+            r.assertBuildStatus(Result.SUCCESS, b);
+            r.assertLogContains("Post-semaphore value is " + conflictingValue, b);
+            r.assertLogContains("a ran on value from b", b);
+        });
+    }
+
+    @Test
+    public void updatedBindingsOnRestart() throws Exception {
+        story.then(r -> {
+            WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+            r.jenkins.getWorkspaceFor(p).child("a.groovy").write("tmp = 'tmp'; { -> tmp}", null);
+            p.setDefinition(new CpsFlowDefinition(
+                    "node() {\n" +
+                    "  a = load('a.groovy')\n" +
+                    "}\n" +
+                    "echo(/before change: ${a()}/)\n" +
+                    "tmp = 'tmp2'\n" +
+                    "echo(/before restart: ${a()}/)\n" +
+                    "semaphore('wait')\n" +
+                    "tmp = 'tmp3'\n" +
+                    "echo(/after restart: ${a()}/)\n", true));
+            WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+            SemaphoreStep.waitForStart("wait/1", b);
+            r.assertLogContains("before change: tmp", b);
+            r.assertLogContains("before restart: tmp2", b);
+        });
+        story.then(r -> {
+            WorkflowJob p = r.jenkins.getItemByFullName("p", WorkflowJob.class);
+            WorkflowRun b = p.getBuildByNumber(1);
+            SemaphoreStep.success("wait/1", null);
+            r.assertBuildStatusSuccess(r.waitForCompletion(b));
+            r.assertLogContains("after restart: tmp3", b);
```

Comment:
With the equality check in `CpsThreadGroup.readResolve` in place, the value here would still be `tmp2`, since the equality check would fail and the bindings would not be linked. Since the bindings are linked before the restart, I think it makes the most sense for them to be linked after the restart too.